opm/common/utility/Serializer.hpp

```diff
@@ -114,5 +132,12 @@ std::string inline Serializer::get<std::string>() {
     return {std::addressof(this->buffer[this->read_pos - length]), length};
 }
 
+template<>
+void inline Serializer::put_vector(const std::vector<bool>& values) {
+    this->put(values.size());
+    for (std::size_t index=0; index < values.size(); index++)
+        this->put(values[index]);
```

Comment:
I am a bit curious about what happens here, since vector of bool returns exotic things from its bracket operator. What is the class that gets put here? I guess only the debugger would tell.

Reply:
I thought the main problem with `std::vector<bool>` was the `.data()` accessor? At least the test with `std::vector<bool>` passes as it should

Reply:
> I am a bit curious about what happens here, since vector of bool returns exotic things from its bracket operator. What is the class that gets put here? I guess only the debugger would tell.

That depends on the exact declaration of the `put` function.  If it is declared as
```C++
ReturnType put(const bool value);
```
then everything is fine (`vector<bool>::reference` is implicitly convertible to `bool`).  If `put()`, on the other hand, is declared as
```C++
template <typename Element>
ReturnType put(const Element& value);
```
then that's a potential problem because `Element` will be deduced to `vector<bool>::reference` and then the semantics and behaviour will depend on the actual function definition/implementation of `put()`.

>  At least the test with `std::vector<bool>` passes as it should

Sadly, this is one of the cases where unit tests aren't sufficient.  A unit test will only exercise the specific details of your particular standard library implementation.

Reply:
I think the concerns are invalid as `values` is const and the called method declared as `const_reference std::vector<bool>::operator[](std::size_t i) const` with `const_reference` being `bool` in the [standard](https://en.cppreference.com/w/cpp/container/vector_bool).

Apart from that I would generally advise against using `std::vector<bool>`. It is only still there to not break backwards compatibility. One should only use it if it memory consumption is really a problem as the interface is surprising and performance is rather bas.

Reply:
> Apart from that I would generally advise against using `std::vector<bool>`

You would rather use e.g `std::vector<int>`?

Reply:
> You would rather use e.g `std::vector<int>`?

Yes, or even `std::vector<char>`. The only downside is that you really should put in a comment saying this is really for storing a boolean etc., but it makes life a lot simpler. (Consider for example that with `vector<bool>` it is not thread-safe to modify different elements, or that there is no guarantee it is even stored contiguously!)

Reply:
I do not want suggesting anything for code I do not know. `std::deque<bool>` might has the least surprises as it is not in disguise.

But maybe there are more important things before refactoring...